src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/loptimizer/ApplyVocabularyMappings.java

```diff
@@ -0,0 +1,70 @@
+package se.liu.ida.hefquin.engine.queryproc.impl.loptimizer;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+import se.liu.ida.hefquin.engine.queryplan.logical.LogicalPlan;
+import se.liu.ida.hefquin.engine.queryplan.logical.LogicalPlanWithNaryRoot;
+import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalOpLocalToGlobal;
+import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalOpMultiwayJoin;
+import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalOpMultiwayUnion;
+import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalOpRequest;
+import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalPlanWithNaryRootImpl;
+import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalPlanWithNullaryRootImpl;
+import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalPlanWithUnaryRootImpl;
+import se.liu.ida.hefquin.engine.queryplan.utils.LogicalOpUtils;
+
+public class ApplyVocabularyMappings implements HeuristicForLogicalOptimization {
+
+	/**
+	 * Rewrites an initial logical plan into a second plan which incorporates translations of local to global vocabulary and request-operator rewriting.
+	 * This method implements the rewriteLogPlan pseudocode of Helgesson's B.Sc thesis.
+	 */
+	@Override
+	public LogicalPlan apply(LogicalPlan inputPlan) {
+		if (inputPlan.getRootOperator() instanceof LogicalOpRequest) {
+			final LogicalOpRequest request = (LogicalOpRequest) inputPlan.getRootOperator();

```

Comment:
That's the line that you say you are not entirely sure about. Generally, it is fine. However, I assume that this line gives you a compiler warning, which you can avoid by better doing it in the following way.
```suggestion
			final LogicalOpRequest<?,?> request = (LogicalOpRequest<?,?>) inputPlan.getRootOperator();
```

Reply:
Additionally, better name this program variable `reqOp` to make clear that this is actually a request operator rather than the request itself. The request you can then get by doing `reqOp.getRequest()`.